pkgs/top-level/octave-packages.nix

```diff
@@ -0,0 +1,269 @@
+
+# Like python-packages.nix, packages from top-level.nix are not in the scope
+# of the `callPackage` used for packages here. So, when we do need packages
+# from outside, we can `inherit` them from `pkgs`.
+{ pkgs
+, lib
+, stdenv
+, fetchurl
+, newScope
+, octave
+}:
+
+with lib;
+
+makeScope newScope (self:
+  let
+    inherit (octave) blas lapack gfortran python texinfo gnuplot jdk;
+
+    callPackage = self.callPackage;
+
+    buildOctavePackage = callPackage ../development/interpreters/octave/build-octave-package.nix {
+      inherit lib stdenv;
+      inherit octave;
+      inherit computeRequiredOctavePackages;
+    };
+
+    wrapOctave = callPackage ../development/interpreters/octave/wrap-octave.nix {
+      inherit octave;
+      inherit (pkgs) makeSetupHook makeWrapper;
+    };
+
+    # Given a list of required Octave package derivations, get a list of
+    # ALL required Octave packages needed for the ones specified to run.
+    computeRequiredOctavePackages = drvs: let
+      # Check whether a derivation is an octave package
+      hasOctavePackage = drv: drv?isOctavePackage;
+      packages = filter hasOctavePackage drvs;
+    in unique (packages ++ concatLists (catAttrs "requiredOctavePackages" packages));
+
+  in {
+
+    inherit callPackage buildOctavePackage computeRequiredOctavePackages;
+
+    inherit (callPackage ../development/interpreters/octave/hooks { })
+      octaveWriteRequiredOctavePackagesHook;
+
+    arduino = callPackage ../development/octave-modules/arduino {
+      inherit (pkgs) arduino;
+      # Full arduino right now. Might be able to use pkgs.arduino-core
+      # Needs arduinoIDE as a runtime dependency.
+    };
+
+    audio = callPackage ../development/octave-modules/audio {
+      rtmidi = pkgs.rtmidi;
+    };
+
+    bim = callPackage ../development/octave-modules/bim { };
+
+    bsltl = callPackage ../development/octave-modules/bsltl { };
+
+    cgi = callPackage ../development/octave-modules/cgi { };
+
+    communications = callPackage ../development/octave-modules/communications {
+      hdf5 = pkgs.hdf5;
+    };
+
+    control = callPackage ../development/octave-modules/control { };
+
+    data-smoothing = callPackage ../development/octave-modules/data-smoothing { };
+
+    database = callPackage ../development/octave-modules/database {
+      postgresql = pkgs.postgresql;
+    };
+
+    dataframe = callPackage ../development/octave-modules/dataframe { };
+
+    dicom = callPackage ../development/octave-modules/dicom {
+      libgdcm = pkgs.gdcm;
+    };
+
+    divand = callPackage ../development/octave-modules/divand { };
+
+    doctest = callPackage ../development/octave-modules/doctest { };
+
+    econometrics = callPackage ../development/octave-modules/econometrics { };
+
+    fem-fenics = callPackage ../development/octave-modules/fem-fenics {
+      # I am still not super sure about this
+      dolfin = python.pkgs.fenics;
```

Comment:
You use python2Packages a bit below so I would suggest to do the same here.

Reply:
The comment was meant more as a reminder for me. When searching through nixpkgs, I noticed that there is no other package that provides `dolfin`. However, it *is* used by python's `fenics` package, which provides python bindings to `fenics`. When I build `python3.pkgs.fenics`, I **do** end up with a derivation and output of `dolfin` which contains the required `dolfin.h` header, but I have no way to get to them.

So right now, the problem is that the only way to access `dolfin` is through the python package, and I cannot access `dolfin` directly.

Reply:
We can copy the header files out of the package into a lib if you like but I am not at all familiar with octave.

Reply:
That would be the only way to get this package (and one other one) to work, since they rely on the fenics package as being stand-alone.

Reply:
"Moving" the library isn't an octave-specific thing. It's definition would be moved to `pkgs/development/libraries` (or another appropriate place), and then the python module can pull it in as a parameter (and perform any additional configuration there through an override). It's just that right now, there is no way to access the `fenics` package/library directly.